pylint/checkers/dataclass_checker.py

```diff
@@ -0,0 +1,137 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Dataclass checkers for Python code."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from astroid import nodes
+
+from pylint.checkers import BaseChecker, utils
+from pylint.interfaces import INFERENCE
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+def _is_dataclasses_module(node: nodes.NodeNG) -> bool:
+    """Utility function to check if node is from dataclasses_module."""
+    return isinstance(node, nodes.Module) and node.name == "dataclasses"
+
+
+def _check_name_or_attrname_eq_to(
+    node: nodes.Name | nodes.Attribute, check_with: str
+) -> Any:
+    """Utility function to check either a Name/Attribute node's name/attrname with a
+    given string.
+    """
+    if isinstance(node, nodes.Name):
+        return node.name == check_with
+    return node.attrname == check_with
+
+
+def _is_annotated_with_dataclass(d_nodes: list[nodes.NodeNG]) -> bool:
+    """Utility function that takes in a list of annotation notes and returns true if we
+    find a node that can be inferred as dataclasses.dataclass.
+    """
+    preprocessed = {
+        n.func if isinstance(n, nodes.Call) else n
+        for n in d_nodes
+        if isinstance(n, (nodes.Name, nodes.Attribute, nodes.Call))
+    }
+    return any(
+        n
+        for n in preprocessed
+        if (isinstance(n, nodes.Name) and n.name == "dataclass")
+        or (
+            isinstance(n, nodes.Attribute)
+            and n.attrname == "dataclass"
+            and _is_dataclasses_module(utils.safe_infer(n.expr))
+        )
+    )
+
+
+class DataclassChecker(BaseChecker):
```

Comment:
Don't we have other `dataclass` specific checks that should go here?

Reply:
I'm not too sure what there is tbh, but I'm following the suggestion of the issue linked to create a new checker for dataclass checks

Reply:
I feel like there are other `dataclass` specific checks in our codebase already that should be moved to this class, but that is something for a later follow up PR.